No backup has been created before trying to reset the current configuration, and removal of the backup was not requested.

<!--
Ensure you have read over [Submitting Issues](https://github.com/chocolatey/.github/blob/main/SUBMITTING-ISSUES.md)

Please check to see if your issue already exists with a quick search of the issues. Start with one relevant term and then add if you get too many results.

NOTE: Keep in mind we have a [Code Of Conduct](https://github.com/chocolatey/.github/blob/main/CODE_OF_CONDUCT.md) that we expect folks to observe when they are looking for support in the Chocolatey community.

Name your issue appropriately: give it a sentence that reads well enough for anyone seeing this in the release notes to what it is.

When writing out the issue details please ensure you are writing it as if you were explaining it to somebody else.
Even if you will be working on and resolving the issue yourself. This helps others to understand the reasons for the
pull request and for it to be searchable in future.
-->

### What You Are Seeing?

When attempting to call `choco upgrade -y sql-server-management-studio` I receive the following error in the logs:

```
2022-11-04 03:19:32,140 4344 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\.chocolatey\sql-server-management-studio.15.0.18424.0".
2022-11-04 03:19:32,156 4344 [DEBUG] - There was no original file at 'C:\ProgramData\chocolatey\.chocolatey\sql-server-management-studio.15.0.18424.0\.registry'
2022-11-04 03:19:32,156 4344 [DEBUG] - There was no original file at 'C:\ProgramData\chocolatey\.chocolatey\sql-server-management-studio.15.0.18424.0\.files'
2022-11-04 03:19:32,172 4344 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\sql-server-management-studio.15.0.18424.0\.extra".
2022-11-04 03:19:32,172 4344 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\sql-server-management-studio.15.0.18424.0\.version".
2022-11-04 03:19:32,172 4344 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\sql-server-management-studio.15.0.18424.0\.sxs".
2022-11-04 03:19:32,172 4344 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\sql-server-management-studio.15.0.18424.0\.pin".
2022-11-04 03:19:32,172 4344 [DEBUG] - Sending message 'HandlePackageResultCompletedMessage' out if there are subscribers...
2022-11-04 03:19:32,172 4344 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\lib\sql-server-management-studio\.chocolateyPending".
2022-11-04 03:19:32,172 4344 [INFO ] -  The upgrade of sql-server-management-studio was successful.
2022-11-04 03:19:32,172 4344 [INFO ] -   Software installed to 'C:\Program Files\Azure Data Studio\'
2022-11-04 03:19:32,172 4344 [DEBUG] - Attempting to delete file "C:\Users\Administrator\AppData\Local\NuGet\Cache\sql-server-management-studio.15.0.18424.0.nupkg".
2022-11-04 03:19:32,187 4344 [WARN ] - 
Chocolatey upgraded 0/0 packages. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
2022-11-04 03:19:32,187 4344 [DEBUG] - Sending message 'PostRunMessage' out if there are subscribers...
2022-11-04 03:19:32,218 4344 [ERROR] - No backup has been created before trying to reset the current configuration, and removal of the backup was not requested.
2022-11-04 03:19:32,218 4344 [ERROR] - More Details: System.InvalidOperationException: No backup has been created before trying to reset the current configuration, and removal of the backup was not requested.
   at chocolatey.infrastructure.app.configuration.ChocolateyConfiguration.reset_config(Boolean removeBackup)
   at chocolatey.infrastructure.app.services.NugetService.upgrade_run(ChocolateyConfiguration config, Action`1 continueAction, Boolean performAction, Action`1 beforeUpgradeAction)
   at chocolatey.infrastructure.app.services.NugetService.upgrade_run(ChocolateyConfiguration config, Action`1 continueAction, Action`1 beforeUpgradeAction)
   at chocolatey.infrastructure.app.services.ChocolateyPackageService.perform_source_runner_function[T](ChocolateyConfiguration config, Func`2 function)
   at chocolatey.infrastructure.app.services.ChocolateyPackageService.upgrade_run(ChocolateyConfiguration config)
   at chocolatey.infrastructure.app.runners.GenericRunner.run(ChocolateyConfiguration config, Container container, Boolean isConsole, Action`1 parseArgs)
   at chocolatey.infrastructure.app.runners.ConsoleApplication.run(String[] args, ChocolateyConfiguration config, Container container)
   at chocolatey.console.Program.Main(String[] args)
2022-11-04 03:19:32,218 4344 [DEBUG] - Exiting with 1
```

### What is Expected?

Not that.

### How Did You Get This To Happen? (Steps to Reproduce)

As described above.

### System Details

* OS Build (In PowerShell run `[System.Environment]::OSVersion.version.tostring()`):  10.0.17763.0 
* Windows PowerShell version (Run: `$PSVersionTable`):
 Name                           Value                                                                                                                                                                                                                                                 
----                           -----                                                                                                                                                                                                                                                 
PSVersion                      5.1.17763.2931                                                                                                                                                                                                                                        
PSEdition                      Desktop                                                                                                                                                                                                                                               
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                                                                               
BuildVersion                   10.0.17763.2931                                                                                                                                                                                                                                       
CLRVersion                     4.0.30319.42000                                                                                                                                                                                                                                       
WSManStackVersion              3.0                                                                                                                                                                                                                                                   
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                                                                   
SerializationVersion           1.1.0.1                                                                                                                                                                                                                                               
* Chocolatey version (Run `choco --version`): 1.2.0

### Output Log

As described above.

--- a/src/chocolatey/infrastructure.app/configuration/ChocolateyConfiguration.cs
+++ b/src/chocolatey/infrastructure.app/configuration/ChocolateyConfiguration.cs
@@ -31,7 +31,7 @@ namespace chocolatey.infrastructure.app.configuration
     public class ChocolateyConfiguration
     {
         [NonSerialized]
-        private ChocolateyConfiguration _originalConfiguration;
+        private Stack<ChocolateyConfiguration> _configurationBackups;
 
         public ChocolateyConfiguration()
         {
@@ -67,9 +67,14 @@ namespace chocolatey.infrastructure.app.configuration
         /// <exception cref="System.Runtime.Serialization.SerializationException">One or more objects in the class or child classes are not serializable.</exception>
         public void start_backup()
         {
+            if (_configurationBackups == null)
+            {
+                _configurationBackups = new Stack<ChocolateyConfiguration>();
+            }
+
             // We do this the easy way to ensure that we have a clean copy
             // of the original configuration file.
-            _originalConfiguration = this.deep_copy();
+            _configurationBackups.Push(this.deep_copy());
         }
 
         /// <summary>
@@ -86,33 +91,42 @@ namespace chocolatey.infrastructure.app.configuration
         /// </remarks>
         public void reset_config(bool removeBackup = false)
         {
-            if (_originalConfiguration == null)
+            if (_configurationBackups == null || _configurationBackups.Count == 0)
             {
                 if (removeBackup)
                 {
                     // If we will also be removing the backup, we do not care if it is already
                     // null or not, as that is the intended state when this method returns.
+                    this.Log().Debug("Requested removal of a configuration backup that does not exist: the backup stack is empty.");
                     return;
                 }
 
                 throw new InvalidOperationException("No backup has been created before trying to reset the current configuration, and removal of the backup was not requested.");
             }
 
+            // Runtime type lookup ensures this also fully works with derived classes (for example: licensed configuration)
+            // without needing to re-implement this method / make it overridable.
             var t = this.GetType();
+
+            var backup = removeBackup ? _configurationBackups.Pop() : _configurationBackups.Peek();
 
             foreach (var property in t.GetProperties(BindingFlags.Public | BindingFlags.Instance))
             {
+                if (property.Name == "_configurationBackups")
+                {
+                    continue;
+                }
+
                 try
                 {
-                    var originalValue = property.GetValue(_originalConfiguration, new object[0]);
+                    var originalValue = property.GetValue(backup, new object[0]);
 
                     if (removeBackup || property.DeclaringType.IsPrimitive || property.DeclaringType.IsValueType || property.DeclaringType == typeof(string))
                     {
                         // If the property is a primitive, a value type or a string, then a copy of the value
                         // will be created by the .NET Runtime automatically, and we do not need to create a deep clone of the value.
                         // Additionally, if we will be removing the backup there is no need to create a deep copy
-                        // for any reference types, as such we also set the reference itself so it is not needed
-                        // to allocate more memory.
+                        // for any reference types. We won't have any duplicate references because the backup is being discarded.
                         property.SetValue(this, originalValue, new object[0]);
                     }
                     else if (originalValue != null)
@@ -130,14 +144,6 @@ namespace chocolatey.infrastructure.app.configuration
                 {
                     throw new ApplicationException("Unable to restore the value for the property '{0}'.".format_with(property.Name), ex);
                 }
-            }
-
-            if (removeBackup)
-            {
-                // It is enough to set the original configuration to null to
-                // allow GC to clean it up the next time it runs on the stored Generation
-                // Heap Table.
-                _originalConfiguration = null;
             }
         }
 

